sdk/rust/src/http.rs

```diff
@@ -30,38 +149,94 @@ pub struct Request {
 /// is no need for streaming bodies.
 pub struct Response {
     /// The status of the response
-    pub status: StatusCode,
+    status: StatusCode,
     /// The response headers
-    pub headers: Vec<(String, Vec<u8>)>,
+    headers: Vec<(String, String)>,
     /// The body of the response as bytes
-    pub body: Vec<u8>,
+    body: Vec<u8>,
 }
 
 impl Response {
     /// Create a new response from a status and optional headers and body
-    pub fn new<S: conversions::IntoStatusCode, B: conversions::IntoBody>(
-        status: S,
-        body: B,
-    ) -> Self {
+    pub fn new(status: impl conversions::IntoStatusCode, body: impl conversions::IntoBody) -> Self {
         Self {
             status: status.into_status_code(),
-            headers: Default::default(),
+            headers: Vec::new(),
             body: body.into_body(),
         }
     }
 
-    /// Create a new response from a status and optional headers and body
-    pub fn new_with_headers<S: conversions::IntoStatusCode, B: conversions::IntoBody>(
-        status: S,
-        headers: Vec<(String, Vec<u8>)>,
-        body: B,
-    ) -> Self {
-        Self {
-            status: status.into_status_code(),
-            headers,
-            body: body.into_body(),
+    /// The response status
+    pub fn status(&self) -> &StatusCode {
+        &self.status
+    }
+
+    /// The response headers
+    ///
+    /// Technically headers do not have to be utf8 encoded but this type assumes they are.
+    /// If you know you'll be dealing with non-utf8 encoded headers, reach more a more powerful
+    /// response type like that found in the `http` crate.
+    pub fn headers(&self) -> &[(String, String)] {
```

Comment:
could we maybe instead introduce a `headers_raw` (and `headers_raw_mut`) that returns `(String, Vec<u8>)`?

I'm imagining that we'd internally store headers as something like `(String, HeaderVal)`, where `HeaderVal` is something like
```rust
enum HeaderVal {
  Bytes(Vec<u8>),
  String(String),
  Both(Vec<u8>, String),
}
```

I bet there's some more optimal representation of this for the case where `Bytes` contains valid unicode, too.

Reply:
I am somewhat worried about this. I think I've managed to [push a change](https://github.com/fermyon/spin/pull/1939/commits/011576ce83b13108d750a19e5c9a656011cd2a02) that supports this without being a maintenance burden, be we do have to get comfortable at some point saying that these types will never cover 100% of use cases and a 3rd party crate might be necessary.